File: crawler/netmarble/netmarble.go
package netmarble

import (
	"context"
	"encoding/json"
	"io/ioutil"
	"log"
	"time"

	"geekermeter-data/crawler"

	"github.com/chromedp/cdproto/cdp"
	"github.com/chromedp/chromedp"
)

var (
	baseURL      = "https://m.netmarble.com/rem/www/noticelist.jsp"
	baseSelector = "#contents > div > div > div > div.recruit_list_wrapper > ul > li >"
)

func Netmarble() []crawler.Job {
	var crawledData []crawler.Job

	ctx, cancel := chromedp.NewContext(context.Background())
	defer cancel()

	var loc string
	err := chromedp.Run(ctx,
		chromedp.Navigate(baseURL),
		chromedp.Location(&loc),
	)
	crawler.ErrHandler(err)

	//find max page
	var pageNode []*cdp.Node
	var totalPage string
	clickerr := chromedp.Run(ctx,
		chromedp.Nodes("#pageCount", &pageNode, chromedp.ByID),
	)
	crawler.ErrHandler(clickerr)
	for _, row := range pageNode {
		totalPage = row.Children[0].NodeValue
	}

	for {
		temp := make([]crawler.Job, 10) // 최소단위 : 10

		chromedp.Sleep(2 * time.Second)
		//url
		var nodes []*cdp.Node
		var titleNode []*cdp.Node
		var newbieNode []*cdp.Node
		var dateNode []*cdp.Node

		err := chromedp.Run(ctx,
			chromedp.Nodes(baseSelector+"div.cw_jopinfo > a", &nodes),
			chromedp.Nodes(baseSelector+"div.cw_jopinfo > a > span.cw_title", &titleNode, chromedp.ByQueryAll),
			chromedp.Nodes(baseSelector+"div.cw_jopinfo > a > span.cw_info > span.cw_type", &newbieNode, chromedp.ByQueryAll),
			chromedp.Nodes(baseSelector+"div.cw_jopinfo > a > span.cw_info > span.cw_range", &dateNode, chromedp.ByQueryAll),
		)
		crawler.ErrHandler(err)

		for i, row := range nodes {
			temp[i].URL = "https://m.netmarble.com/rem/www" + row.AttributeValue("href")[1:]
			temp[i].Origin = "Netmarble"
		}
		//title
		for i, row := range titleNode {
			temp[i].Title = row.Children[0].NodeValue
		}
		for i, row := range newbieNode {
			temp[i].Newbie = row.Children[0].NodeValue
		}
		for i, row := range dateNode {
			temp[i].StartDate = row.Children[0].NodeValue[:8]
		}

		crawledData = append(crawledData, temp...)

		var currPage string
		chromedp.Sleep(2 * time.Second)
		clickerr := chromedp.Run(ctx,
			chromedp.Text("#contents > div > div > div > div.recruit_list_wrapper > div.recruit_pagination > span.page_current", &currPage),
		)
		if clickerr != nil {
			log.Fatal(clickerr)
		}

		if currPage == totalPage {
			break
		} else {
			err = chromedp.Run(ctx,
				chromedp.Click("#contents > div > div > div > div.recruit_list_wrapper > div.recruit_pagination > button.page_next", chromedp.NodeVisible),
			)

		}

		//fmt.Println(i)
	}
	for _, dat := range crawledData {
		log.Printf("%s", dat.Title)
		log.Printf("%s", dat.URL)
	}

	return crawledData
}

func BodyText(box crawler.Job) { //현재 쓸데없는 값까지 하는 중->예외처리 실패로 인해..
	ctx, cancel := chromedp.NewContext(context.Background())
	defer cancel()

	// run task list
	var loc string
	err := chromedp.Run(ctx,
		chromedp.Navigate(box.URL),
		chromedp.Location(&loc),
	)
	crawler.ErrHandler(err)

	var nodes []*cdp.Node
	var candNodes []*cdp.Node
	var sectionName string
	err = chromedp.Run(ctx,
		chromedp.Nodes("body > div.main > div.position-show > div > div > div.content-body.col-item.col-xs-12.col-sm-12.col-md-12.col-lg-8>section",
			&nodes),
	)
	crawler.ErrHandler(err)
	for _, row := range candNodes {
		next_html := row.FullXPath() + "/td[2]"
		var temp string
		err = chromedp.Run(ctx,
			chromedp.Text(next_html,
				&temp),
		)
		if temp == "기간" {
			err = chromedp.Run(ctx,
				chromedp.Text(row.FullXPath()+"/td[3]",
					&box.StartDate),
			)
		}

		crawler.ErrHandler(err)
	}
	acceptList := []string{"section-stacks", "section-position", "section-requirements", "section-preference"}
	presentList := make([]string, 4)
	presentNum := 0
	for _, row := range nodes {
		sectionName = row.AttributeValue("class")
		_, found := Find(acceptList, sectionName)
		if found == true {
			presentList[presentNum] = sectionName
			presentNum++
		}
	}
	presentList = presentList[:presentNum]

	box.Content = make([]string, presentNum)
	for index, name := range presentList {
		selurl := "body > div.main > div.position-show > div > div > div.content-body.col-item.col-xs-12.col-sm-12.col-md-12.col-lg-8 > section." + name
		box.Content[index] = getdetail(ctx, selurl)
	}

	doc, _ := json.Marshal(box)
	_ = ioutil.WriteFile("./dataset/test/"+crawler.Exceptspecial(box.URL)+".json", doc, 0644)

}

File: crawler/netmarble/netmarble.py
import re, os, sys
import selenium
sys.path.append(os.path.dirname(os.path.abspath(os.path.dirname(__file__))))
from util import *
from selenium import webdriver
from bs4 import BeautifulSoup
import warnings
warnings.filterwarnings(action='ignore')


def netmarble(driver_path=None):
    if not driver_path:
        driver_path='/Users/mingihong/chromedriver'
    driver = headless(driver_path)
#    driver = webdriver.Chrome(driver_path)
    driver.get('https://m.netmarble.com/rem/www/noticelist.jsp')

    final_data = []
    while True:
        crawled_data = []
        WebDriverWait(driver, 10).until(
            EC.presence_of_element_located((By.CSS_SELECTOR, "#contents > div > div > div > div.recruit_list_wrapper > ul"))
        )
        html = driver.page_source
        soup = BeautifulSoup(html,'html.parser')

        posts = soup.select('#contents > div > div > div > div.recruit_list_wrapper > ul > li')
        for post in posts:

            post_date = transfrom_date(post.select('div.cw_jopinfo > a > span.cw_info > span.cw_range')[0].text)
            if not post_date:
                continue
            post_title = post.select('div.cw_jopinfo > a > span.cw_title')[0].text
            post_url = 'https://m.netmarble.com/rem/www'+post.select('div.cw_jopinfo > a')[0].get('href')[1:]
            post_newbie = make_newbie(post.select('div.cw_jopinfo > a > span.cw_info > span.cw_type')[0].text)

            crawled_data.append(
                job(post_title,post_url,'netmarble',post_date,post_newbie).data
            )

        final_data.extend(crawled_data)
        _next = driver.find_element_by_css_selector('#contents > div > div > div > div.recruit_list_wrapper > div.recruit_pagination > button.page_next')
        if not _next.get_attribute('disabled'):
            _next.click()
        else:
            break


    return final_data, driver

def body_text(driver,json):
    driver.get(json['url'])

    html = driver.page_source
    soup = BeautifulSoup(html,'html.parser')
    txt = soup.select('#tmpCapture > div > table')
    if txt:
        json['contents'] = txt[0].text
        return json
    else:
        return None


def run(driver_path=None):

    print('start crawling : netmarble')
    json_list, driver = netmarble(driver_path)
    print('The number of netmarble post : %d'%(len(json_list)))
    for i,json in enumerate(json_list):
        if i%10 == 0:
            print('netmarble post : %d'%(i))
        json = body_text(driver,json)
    driver.quit()
    print('finish crawling : netmarble')

    return json_list


if __name__ == "__main__":
    run()
